### DISCUSSION TITLE
How to pull a remote branch into a local one

### DISCUSSION THREAD

**Discussion by dbatten5** [NeogitOrg/neogit#1344] [Q&A]
Sure this is a very basic question but for the life of me I can't see a quick way to to do it. What's the best way to pull the master/main branch from the remote to a local one? I've tried the pull from elsewhere option and then selecting master but I get this error 

>The coroutine failed with this message: ...hare/nvim/lazy/neogit/lua/neogit/popups/pull/actions.lua:17: attempt to concatenate local 'remote' (a nil value)

Maybe I'm missing some setup there. In any case though it's a bit of a drag going through the process of `:Neogit` followed by `p` followed by `e` then finding the master branch in the list. Is there a quicker way or shortcut to doing this?

Note that I have 
```
[pull]
	rebase = true
```
in my git config

**Comment by CKolkey** [comment]
Ahh, so I think I just didn't consider this case (pulling from a local branch) because that's the same as merge. Local branches shouldn't be an option on that popup - I'll fix that. 

Anyways, to do what you are trying to do, you can do `:Neogit merge` `m` or `:Neogit` `m` `m`

**Reply by dbatten5** [reply]
Thanks for the quick response @CKolkey ! Would `:Neogit` `m` `m` be merging the remote master branch or the local one?

**Reply by CKolkey** [reply]
Oh! You said from the remote. Sorry. Just pick `origin/master` from the list of branches.

**Reply by CKolkey** [reply]
So `:Neogit` `p` `e` `origin/master`

**Reply by CKolkey** [reply]
(you could also use `merge` and choose `origin/master`)

**Reply by dbatten5** [reply]
Ahh yes that's the one, sorry should have realised `origin/master` was the one to go for there. Although that has returned an error
```
git name-rev --name-only --no-undefined --refs=refs/heads/* --exclude=*/HEAD --exclude=*/refs/heads/* 851f451c979084da37d6fcc9f97161ef20ca61d6:

    fatal: cannot describe '851f451c979084da37d6fcc9f97161ef20ca61d6'
```

**Reply by CKolkey** [reply]
And if you have `master` checked out locally, then you're just pulling from the upstream, so that's `pu`

**Reply by CKolkey** [reply]
Weird... if you fetch the upstream does it still do that?

**Reply by dbatten5** [reply]
Yeah same error. Weirdly it doesn't seem to affect the actual actions taken, I can see I've pulled from the latest remote master

**Reply by CKolkey** [reply]
Looks like the only place I call "name-rev" is if neogit thinks you're in the middle of a rebase...

Does your `.git/` dir have a `rebase-merge/onto` or `rebase-apply/onto` file?

https://github.com/NeogitOrg/neogit/blob/2d749c1236bae5d4513b87276083a3e5d7044962/lua/neogit/lib/git/rebase.lua#L188

**Reply by dbatten5** [reply]
Yep so I do have 
```
[pull]
    rebase = true
```
in my git config and for this particular pull from master I had a couple of conflicts to sort out during the rebase. Might that be the source of the issue?

**Reply by CKolkey** [reply]
interesting, yeah, it might be. If there are conflicts then the rebase wouldn't resolve, and when we re-fetch the state of the repo it'll be mid-rebase.

If you do `git log` with that hash, does it work?

**Reply by dbatten5** [reply]
Yeah it does work, that hash is on the remote master branch (and my local one now after the rebase)

**Reply by CKolkey** [reply]
Ok! If you wanna try this branch out and let me know if it works, I'd appreciate it :) https://github.com/NeogitOrg/neogit/pull/1346

**Reply by CKolkey** [reply]
![Screenshot 2024-05-31 at 16 03 46](https://github.com/NeogitOrg/neogit/assets/7228095/19db3b86-4836-4c5e-ae9f-6c449a0d6c38)
You know what, I can reproduce the issue, and will just fix it ;)

**Reply by dbatten5** [reply]
Awesome, thanks so much, and thanks for all your hard work on this awesome plugin
